### PR TITLE
Adds filesystem

### PR DESCRIPTION
**Pull request by aidan-casey**
Closes tempestphp/tempest-framework#272 when merged.

### CODE REVIEW

**Comment by brendt** [comment]
> when merged.

Can I?? 😁

Left some comments, only afterwards I noticed it was still draft 🫣

File: src/Filesystem/Exception/FileAlreadyExists.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Tempest\Filesystem\Exception;
+
+use RuntimeException;
+
+final class FileAlreadyExists extends RuntimeException
+{
+    public static function atPath(string $path): self
+    {
+        return new self();
```
Comment (brendt) [inline src/Filesystem/Exception/FileAlreadyExists.php:13]: You're not doing anything with `$path`?

File: src/Filesystem/Exception/UnableToCreateDirectory.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+declare(strict_types=1);
+
+namespace Tempest\Filesystem\Exception;
+
+use RuntimeException;
+
+class UnableToCreateDirectory extends RuntimeException
+{
+    public static function atLocation(string $location, string $because): self
```
Comment (brendt) [inline src/Filesystem/Exception/UnableToCreateDirectory.php:11]: What's the difference between `location` and `path` in the previous exception?

File: src/Filesystem/FilesystemFactory.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Tempest\Filesystem;
+
+use Tempest\Filesystem\Local\LocalDriver;
+
+final class FilesystemFactory
+{
+    // TODO: We may want to get rid of this,
+    // just playing around with syntax.
+    public static function local(string $basePath): Filesystem
```
Comment (brendt) [inline src/Filesystem/FilesystemFactory.php:13]: Yeah I don't see much benefit from this class

File: src/Filesystem/Local/LocalDriver.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            directory: $location,
+            permissions: $permissions,
+            recursive: true
+        );
+
+        if ($directoryWasCreated) {
+            return;
+        }
+
+        throw UnableToCreateDirectory::atLocation(
```
Comment (brendt) [inline src/Filesystem/Local/LocalDriver.php:68]: I'd move this exception to the `if` block instead. I prefer the happy path to be the "primary path"

File: src/Filesystem/Local/LocalDriver.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        error_clear_last();
+
+        $directoryWasDeleted = @rmdir($location);
+
+        if ($directoryWasDeleted) {
+            return;
+        }
+
+        throw UnableToDeleteDirectory::atLocation(
```
Comment (brendt) [inline src/Filesystem/Local/LocalDriver.php:84]: Same with this one

File: tests/Unit/Support/Fixtures/test_file.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+Hi. I'm here for testing.
```
Comment (brendt) [inline tests/Unit/Support/Fixtures/test_file.txt:1]: 👋

File: src/Filesystem/Stream.php [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Filesystem;
+
+interface Stream
```
Comment (brendt) [inline src/Filesystem/Stream.php:7]: Is a stream used somewhere? Or is it just an alternative provided by the framework?

**Review by brendt — CHANGES_REQUESTED** [review]
Overall, great PR. Just a couple minor details I'd change

File: src/Filesystem/Driver/LocalFilesystemDriver.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    public function createDirectory(string $path, int $mode): void
+    {
+        // TODO: Error handling.
+        mkdir($path, $mode, true);
+        clearstatcache(false, $path);
+    }
+
+    public function deleteDirectory(string $path): void
+    {
+        // TODO: Error handling; recursive?
```
[the remark this answers is not in the document]
Comment (aidan-casey) [reply src/Filesystem/Driver/LocalFilesystemDriver.php]: It's a draft. ;)

Recursive delete is indeed on my list of things to implement.

File: src/Filesystem/FilesystemDriver.php [hunk 9]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Filesystem;
+
+interface FilesystemDriver
```
[the remark this answers is not in the document]
Comment (aidan-casey) [reply src/Filesystem/FilesystemDriver.php]: Yeah, I was going to add something like `createFile` or `touchFile`. This interface is by no means complete for the overall PR.

File: src/Filesystem/Exception/FileAlreadyExists.php (hunk 1 above)
Comment (aidan-casey) [reply src/Filesystem/Exception/FileAlreadyExists.php:13]: Not _yet_. 😉

File: src/Filesystem/Exception/UnableToCreateDirectory.php (hunk 2 above)
Comment (aidan-casey) [reply src/Filesystem/Exception/UnableToCreateDirectory.php:11]: `$path` is where I started, `$location` is what I am refactoring to.

File: src/Filesystem/Local/LocalDriver.php (hunk 4 above)
Comment (aidan-casey) [reply src/Filesystem/Local/LocalDriver.php:68]: 🥲

File: src/Filesystem/Local/LocalDriver.php (hunk 5 above)
Comment (aidan-casey) [reply src/Filesystem/Local/LocalDriver.php:84]: 🥲

File: tests/Unit/Support/Fixtures/test_file.txt (hunk 6 above)
Comment (aidan-casey) [reply tests/Unit/Support/Fixtures/test_file.txt:1]: 👋

File: src/Filesystem/Stream.php (hunk 7 above)
Comment (aidan-casey) [reply src/Filesystem/Stream.php:7]: A stream interface is returned by `createStream` on the Filesystem driver so you could create a file stream that is filesystem agnostic.

Comment (brendt) [reply src/Filesystem/Stream.php:7]: Oh, I missed that, I think that's great :)

**Comment by brendt** [comment]
@aidan-casey wdyt we're going to do with this? In the current architecture it should be moved to tempest/core and probably needs some more finetuning. We could also close it for now and revisit it at a later point. I don't have a strong opinion either way.

**Comment by brendt** [comment]
I'm gonna close this one. Let's reconsider at a later point

**Pull request closed without merging**